projects/js-packages/critical-css-gen/src/browser-interface-playwright.ts

```diff
@@ -0,0 +1,125 @@
+		super();
+	}
+
+	private async getTabs() {
+		if ( typeof this.tabs === 'undefined' ) {
+			await this.openUrls( this.context, this.urls );
+		}
+
+		return this.tabs;
+	}
+
+	/**
+	 * Open an array of urls in a new browser context.
+	 *
+	 * Take a browser instance and an array of urls to open in new tabs.
+	 *
+	 * @param {BrowserContext} context - Browser context to use.
+	 * @param {string[]}       urls    - Array of urls to open.
+	 * @returns {Promise< TabsByUrl >} Promise resolving to the browser context.
+	 */
+	private async openUrls( context: BrowserContext, urls: string[] ): Promise< void > {
+		this.tabs = await objectPromiseAll< Tab >(
+			urls.reduce( ( set, url ) => {
+				set[ url ] = this.newTab( context, url );
+				return set;
+			}, {} )
+		);
+	}
+
+	/**
+	 * Open url in a new tab in a given browserContext.
+	 *
+	 * @param {BrowserContext} browserContext - Browser context to use.
+	 * @param {string}         url            - Url to open.
+	 * @returns {Promise<Page>} Promise resolving to the page instance.
+	 */
+	private async newTab( browserContext: BrowserContext, url: string ): Promise< Tab > {
+		const tab = {
+			page: await browserContext.newPage(),
+			statusCode: null,
+		};
+		tab.page.on( 'response', async response => {
+			if ( response.url() === url ) {
+				tab.statusCode = response.status();
+			}
+		} );
+
+		await tab.page.goto( url, { timeout: PAGE_GOTO_TIMEOUT_MS } );
+
+		return tab;
+	}
+
+	async runInPage< ReturnType >(
+		pageUrl: string,
+		viewport: Viewport | null,
+		method: BrowserRunnable< ReturnType >,
+		...args: unknown[]
+	): Promise< ReturnType > {
+		const tabs = await this.getTabs();
+		const tab = tabs[ pageUrl ];
+
+		if ( ! tab || ! tab.page ) {
+			throw new Error( `Playwright interface does not include URL ${ pageUrl }` );
+		}
+
+		// Bail early if the page returned a non-200 status code.
+		if ( ! tab.statusCode || ! this.isOkStatus( tab.statusCode ) ) {
+			const error = new HttpError( { url: pageUrl, code: tab.statusCode } );
+			this.trackUrlError( pageUrl, error );
+			throw error;
+		}
+
+		if ( viewport ) {
+			await tab.page.setViewportSize( viewport );
+		}
+
+		// The inner window in Playwright is the directly accessible main window object.
+		// The evaluating method does not need a separate window object.
+		// Call inner method within the Playwright context.
+		return tab.page.evaluate( method, { innerWindow: null, args } );
+	}
+
+	/**
+	 * Replacement for browser.fetch, uses node-fetch to simulate the same
+	 * interface.
+	 *
+	 * @param {string} url     - URL to fetch.
+	 * @param {object} options - Fetch options.
+	 * @param {string} _role   - 'css' or 'html' indicating what kind of thing is being fetched.
+	 * @returns {Promise<Response>} A promise that resolves to the fetch response.
+	 */
+	async fetch( url: string, options: FetchOptions, _role: 'css' | 'html' ) {
+		const nodeFetch = await import( 'node-fetch' );
+
+		return nodeFetch.default( url, options );
```

Comment:
I wonder if this can be replaced with [the native implementation of `fetch()`](https://nodejs.org/api/globals.html#fetch) that has been enabled by default since Node 18? That would let us drop the `node-fetch` dependency.

Reply:
I love this. I have no idea how it didn't click for me.

The Boost Cloud (that's using the node build of this package), is running on version 18.8.0, so it's a no-brainer...

Thank you so much for this suggestion!